src/control/cmd/daos_server/storage_legacy.go

```diff
@@ -0,0 +1,216 @@
+	DisableVFIO  bool   `long:"disable-vfio" description:"Force SPDK to use the UIO driver for NVMe device access"`
+}
+
+// Validate checks both only options are not set and returns flags to direct
+// which subsystem types to prepare.
+func (cmd *legacyPrepCmd) Validate() (bool, bool, error) {
+	prepScm := cmd.ScmOnly || !cmd.NvmeOnly
+	if prepScm {
+		cmd.Debug("legacy prepare of scm selected")
+	}
+	prepNvme := cmd.NvmeOnly || !cmd.ScmOnly
+	if prepNvme {
+		cmd.Debug("legacy prepare of nvme selected")
+	}
+
+	if cmd.ScmOnly && cmd.NvmeOnly {
+		return false, false, errors.New(
+			"scm-only and nvme-only options should not be set together")
+	}
+
+	return prepScm, prepNvme, nil
+}
+
+func (cmd *legacyPrepCmd) prep(scs *server.StorageControlService, iommuEnabled bool) error {
+	var errSCM, errNVMe error
+
+	doSCM, doNVMe, err := cmd.Validate()
+	if err != nil {
+		return err
+	}
+
+	if cmd.Reset {
+		if doSCM {
+			var rnc removeNamespacesCmd
+			if err := convert.Types(cmd, &rnc); err != nil {
+				return errors.Wrap(err, "converting legacy prepare command")
+			}
+			rnc.LogCmd = cmdutil.LogCmd{Logger: cmd.Logger}
+			errSCM = rnc.resetPMem(scs.ScmPrepare)
+		}
+		if doNVMe {
+			var rdc releaseDrivesCmd
+			if err := convert.Types(cmd, &rdc); err != nil {
+				return errors.Wrap(err, "converting legacy prepare command")
+			}
+			rdc.LogCmd = cmdutil.LogCmd{Logger: cmd.Logger}
+			errNVMe = rdc.resetNVMe(scs.NvmePrepare, iommuEnabled)
+		}
+	} else {
+		if doSCM {
+			var cnc createNamespacesCmd
+			if err := convert.Types(cmd, &cnc); err != nil {
+				return errors.Wrap(err, "converting legacy prepare command")
+			}
+			cnc.LogCmd = cmdutil.LogCmd{Logger: cmd.Logger}
+			errSCM = cnc.preparePMem(scs.ScmPrepare)
+		}
+		if doNVMe {
+			var pdc prepareDrivesCmd
+			if err := convert.Types(cmd, &pdc); err != nil {
+				return errors.Wrap(err, "converting legacy prepare command")
+			}
+			pdc.LogCmd = cmdutil.LogCmd{Logger: cmd.Logger}
+			errNVMe = pdc.prepareNVMe(scs.NvmePrepare, iommuEnabled)
+		}
+	}
+
+	switch {
+	case errSCM != nil && errNVMe != nil:
+		return errors.Wrap(common.ConcatErrors([]error{errSCM, errNVMe}, nil),
+			"storage prepare command returned multiple errors")
+	case errSCM != nil:
+		return errSCM
+	}
+
+	return errNVMe
+}
+
+func (cmd *legacyPrepCmd) Execute(args []string) error {
+	cmd.Info("storage prepare subcommand is deprecated, use nvme or pmem subcommands instead")
+
+	// This is a little ugly, but allows for easier unit testing.
+	// FIXME: With the benefit of hindsight, it seems apparent
+	// that we should have made these Execute() methods thin
+	// wrappers around more easily-testable functions.
+	if cmd.scs == nil {
+		cmd.scs = server.NewStorageControlService(cmd.Logger, config.DefaultServer().Engines)
+	}
+
+	iommuEnabled, err := hwprov.DefaultIOMMUDetector(cmd).IsIOMMUEnabled()
+	if err != nil {
+		return errors.Wrap(err, "detecting if iommu is enabled")
+	}
```

Comment:
that would invalidate the way the unit tests work

Reply:
Maybe the cmd could have a hardware.IOMMUDetector member that could be called. That would allow the tests to inject the value while having the interface of the prep() function be a bit cleaner.

Reply:
done